exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	for _, metricData := range getMetricData(metrics) {
+		if metricData.valueType == prometheus.UntypedValue {
+			m, err := prometheus.NewConstHistogram(metricData.description, metricData.histogramCount, metricData.histogramSum, metricData.histogramBuckets, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		} else {
+			m, err := prometheus.NewConstMetric(metricData.description, metricData.valueType, metricData.value, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		}
+	}
+}
+
+// metricData holds the metadata as well as values for individual data points.
+type metricData struct {
+	name             string
+	description      *prometheus.Desc
+	attributeValues  []string
+	valueType        prometheus.ValueType
+	value            float64
+	histogramCount   uint64
+	histogramSum     float64
+	histogramBuckets map[float64]uint64
+}
+
+func getMetricData(metrics metricdata.ResourceMetrics) []*metricData {
+	allMetrics := make([]*metricData, 0)
+	for _, scopeMetrics := range metrics.ScopeMetrics {
+		for _, m := range scopeMetrics.Metrics {
+			switch v := m.Data.(type) {
+			case metricdata.Histogram:
+				allMetrics = append(allMetrics, getHistogramMetricData(v, m)...)
+			case metricdata.Sum[int64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Sum[float64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Gauge[int64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			case metricdata.Gauge[float64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			}
+		}
+	}
+
+	return allMetrics
+}
+
+func getHistogramMetricData(histogram metricdata.Histogram, m metricdata.Metrics) []*metricData {
+	dataPoints := make([]*metricData, 0)
+	for _, dp := range histogram.DataPoints {
+		keys, values := getAttrs(dp.Attributes)
+		desc := prometheus.NewDesc(m.Name, m.Description, keys, nil)
+		buckets := make(map[float64]uint64)
+		for i, bound := range dp.Bounds {
+			buckets[bound] = dp.BucketCounts[i]
+		}
+		md := &metricData{
+			name:             m.Name,
+			description:      desc,
+			attributeValues:  values,
+			valueType:        prometheus.UntypedValue,
+			histogramCount:   dp.Count,
+			histogramSum:     dp.Sum,
+			histogramBuckets: buckets,
+		}
+		dataPoints = append(dataPoints, md)
+	}
+	return dataPoints
+}
+
+func getSumMetricData[N int64 | float64](sum metricdata.Sum[N], m metricdata.Metrics) []*metricData {
+	dataPoints := make([]*metricData, 0)
+	for _, dp := range sum.DataPoints {
+		keys, values := getAttrs(dp.Attributes)
+		desc := prometheus.NewDesc(m.Name, m.Description, keys, nil)
+		md := &metricData{
+			name:            m.Name,
+			description:     desc,
+			attributeValues: values,
+			valueType:       prometheus.CounterValue,
```

Comment:
Non-monotonic sums are converted to prometheus Gauges.

If the aggregation temporality is delta, it should be dropped (for now.  Add a TODO to convert to cumulative temporality)

Source: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#sums-1